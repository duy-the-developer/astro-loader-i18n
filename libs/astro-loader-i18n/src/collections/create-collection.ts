import { i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { z } from "astro/zod";

type Options = {
  locales: string[];
  routePattern: string;
};

export function createCollection(options: Options): z.infer<typeof i18nLoaderSchema>[] {
  const { locales, routePattern } = options;
  return locales.map((locale) => ({ locale: locale, translationId: routePattern }));
}
