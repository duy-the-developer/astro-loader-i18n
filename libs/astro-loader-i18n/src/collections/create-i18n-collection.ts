import { i18nLoaderSchema } from "../schemas/i18n-loader-schema";
import { z } from "astro/zod";

type Options = {
  locales: string[];
  routePattern: string;
};

export type I18nCollection = { data: z.infer<typeof i18nLoaderSchema> }[];

export function createI18nCollection(options: Options): I18nCollection {
  const { locales, routePattern } = options;

  return locales.map((locale) => ({ data: { locale: locale, translationId: routePattern, contentPath: "" } }));
}
