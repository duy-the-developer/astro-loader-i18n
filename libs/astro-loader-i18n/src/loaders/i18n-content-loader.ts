import { glob, Loader, LoaderContext } from "astro/loaders";
import { pruneLocales, getAllUniqueKeys } from "../utils/collection";
import { createTranslationId } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nContentLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-content-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseData = context.parseData;
      const parseDataProxy: typeof parseData = (props) => {
        if (!props.filePath) return parseData(props);
        const locale = defaultLocale;
        const translationId = createTranslationId(props.filePath);
        return parseData({ ...props, data: { ...props.data, locale, translationId } });
      };
      context.parseData = parseDataProxy;

      await globLoader.load(context);

      const entries = context.store.entries();
      context.store.clear();

      entries.forEach(([, entry]) => {
        const entryLocales = Array.from(getAllUniqueKeys(entry.data)).filter((key) => localeCodes.includes(key));
        if (entryLocales.length === 0) {
          context.store.set(entry);
        } else {
          entryLocales.forEach((locale) => {
            const entryData = pruneLocales(entry.data, entryLocales, locale);
            context.store.set({ ...entry, id: `${entry.id}/${locale}`, data: { ...entryData, locale } });
          });
        }
      });
    },
  };
}
