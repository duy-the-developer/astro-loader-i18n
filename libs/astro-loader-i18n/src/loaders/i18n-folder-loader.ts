import { glob, Loader, LoaderContext } from "astro/loaders";
import { createCommonTranslationId, parseLocale } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nFolderLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {
      await globLoader.load(context);
      const entries = context.store.values();

      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      context.store.clear();
      entries.forEach((entry) => {
        if (!entry.filePath) return;
        const locale = parseLocale(entry.filePath, localeCodes, defaultLocale);

        const newEntry = {
          ...entry,
          data: {
            ...entry.data,
            locale,
            commonTranslationId: createCommonTranslationId(entry.filePath, locale),
          },
        };
        context.store.set(newEntry);
      });
    },
  };
}
