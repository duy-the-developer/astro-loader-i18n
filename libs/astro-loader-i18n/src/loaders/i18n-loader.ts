import { glob, Loader, LoaderContext } from "astro/loaders";
import { createTranslationId, parseLocale } from "../utils/path";

type GlobOptions = Parameters<typeof glob>[0];

export function i18nLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseDataProxy: typeof context.parseData = (props) => {
        if (!props.filePath) return context.parseData(props);
        const locale = parseLocale(props.filePath, localeCodes, defaultLocale);
        const translationId = createTranslationId(props.filePath, locale);
        return context.parseData({ ...props, data: { ...props.data, locale, translationId } });
      };

      const globContext = Object.assign({}, context, { parseData: parseDataProxy });
      await globLoader.load(globContext);
    },
  };
}
