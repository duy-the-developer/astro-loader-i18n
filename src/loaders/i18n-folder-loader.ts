import { glob, Loader, LoaderContext } from "astro/loaders";
import { GlobOptions } from "../types";

export function i18nFolderLoader(options: GlobOptions ): Loader {
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {
      await glob(options).load(context);

    }
  }
}
