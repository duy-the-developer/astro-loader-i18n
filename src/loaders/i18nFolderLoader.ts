import { Loader, LoaderContext } from "astro/loaders";

export function i18nFolderLoader(): Loader {
  return {
    name: "i18n-folder-loader",
    load: async (context: LoaderContext): Promise<void> => {}
  }
}
