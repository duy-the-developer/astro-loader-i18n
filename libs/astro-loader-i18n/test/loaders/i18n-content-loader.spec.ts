import { beforeEach, describe, expect, it, vi } from "vitest";
import { LoaderContext } from "astro/loaders";
import { createLoaderContext } from "../__mocks__/loader-context";
import { contentCollectionFixture } from "../__fixtures__/collections";
import { i18nContentLoader } from "../../src/loaders/i18n-content-loader";

vi.mock("astro/loaders", () => {
  return {
    glob: () => {
      return {
        load: vi.fn().mockImplementation(async (context: LoaderContext) => {
          contentCollectionFixture.forEach(async (entry) => {
            context.store.set({ ...entry, data: await context.parseData(entry) });
          });
        }),
      };
    },
  };
});

describe("i18nContentLoader", () => {
  let context: LoaderContext;

  beforeEach(() => {
    context = createLoaderContext();
  });

  it("should put common translation id and locale in data", async () => {
    const loader = i18nContentLoader({ pattern: "**/*.yml", base: "./src/content/gallery" });
    await loader.load(context);

    const entries = context.store.entries();
    expect(entries).toMatchSnapshot();
  });
});
