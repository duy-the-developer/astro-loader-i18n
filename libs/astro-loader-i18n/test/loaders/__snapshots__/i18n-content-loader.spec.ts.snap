// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nContentLoader > should put common translation id and locale in data 1`] = `
[
  [
    "space",
    {
      "data": {
        "cover": "./space1.jpg",
        "images": [
          {
            "src": "./space1.jpg",
          },
          {
            "src": "./space2.jpg",
          },
          {
            "src": "./space3.jpg",
          },
          {
            "src": "./space4.jpg",
          },
          {
            "src": "./space5.jpg",
          },
        ],
        "locale": "zh-CN",
        "title": "太空",
        "translationId": "src/content/gallery/space.yml",
      },
      "filePath": "src/content/gallery/space.yml",
      "id": "space",
    },
  ],
  [
    "nature",
    {
      "data": {
        "cover": "./nature1.jpg",
        "images": [
          {
            "src": "./nature1.jpg",
          },
          {
            "src": "./nature2.jpg",
          },
          {
            "src": "./nature3.jpg",
          },
          {
            "src": "./nature4.jpg",
          },
          {
            "src": "./nature5.jpg",
          },
        ],
        "locale": "zh-CN",
        "title": "自然",
        "translationId": "src/content/gallery/nature.yml",
      },
      "filePath": "src/content/gallery/nature.yml",
      "id": "nature",
    },
  ],
  [
    "animals",
    {
      "data": {
        "cover": "./animals1.jpg",
        "images": [
          {
            "src": "./animals1.jpg",
          },
          {
            "src": "./animals2.jpg",
          },
          {
            "src": "./animals3.jpg",
          },
          {
            "src": "./animals4.jpg",
          },
          {
            "src": "./animals5.jpg",
          },
        ],
        "locale": "zh-CN",
        "title": "动物",
        "translationId": "src/content/gallery/animals.yml",
      },
      "filePath": "src/content/gallery/animals.yml",
      "id": "animals",
    },
  ],
]
`;
